File: Bwl.Imaging/Geometry/Region/RegionWithVector.vb
Imports System.Drawing

Public Class RegionWithVector
    Inherits Region
    Implements ICloneable

    Public Shared ReadOnly Property VectorMarker As String = "Vector"

    Public ReadOnly Property VectorCaption As String
        Get
            Return Caption + VectorMarker
        End Get
    End Property

    Public ReadOnly Property VectorID As String
        Get
            Return ID + VectorMarker
        End Get
    End Property

    Public Property Vector As New PointF()

    Public Property ImagingVector As Vector
        Get
            Dim point1 = New PointF(Points.Average(Function(item) item.X), Points.Average(Function(item) item.Y))
            Dim point2X = point1.X + Vector.X : point2X = If(point2X < 0, 0, point2X) : point2X = If(point2X > 1, 1, point2X)
            Dim point2Y = point1.Y + Vector.Y : point2Y = If(point2Y < 0, 0, point2Y) : point2Y = If(point2Y > 1, 1, point2Y)
            Dim point2 = New PointF(point2X, point2Y)
            Return New Vector(point1, point2)
        End Get
        Set(value As Vector)
            Vector = New PointF(value.Points(1).X - value.Points(0).X, value.Points(1).Y - value.Points(0).Y)
        End Set
    End Property

    Public Sub New()
        MyBase.New(Guid.NewGuid.ToString(), {New PointF(), New PointF(), New PointF(), New PointF()}) 'Как Tetragon
    End Sub

    Public Sub New(points As PointF())
        MyBase.New(Guid.NewGuid.ToString(), points)
    End Sub

    Public Sub New(id As String)
        MyBase.New(id, {New PointF(), New PointF(), New PointF(), New PointF()}) 'Как Tetragon
    End Sub

    Public Sub New(id As String, points As PointF())
        MyBase.New(id, points) 'Как Tetragon
    End Sub

    Public Overrides Function ToDisplayObjects(Optional fullDisplay As Boolean = True, Optional textSizeF As Single = 0.013,
                                               Optional channelIdxKey As String = "{ChannelIdxKey}") As DisplayObject()
        Dim displayObjects = New List(Of DisplayObject)(MyBase.ToDisplayObjects(fullDisplay, textSizeF, channelIdxKey))
        'Vector
        If Vector.X <> 0 Or Vector.Y <> 0 Then
            Dim displayObjectVector = New DisplayObject(VectorID, Color, ImagingVector) With
            {
                .Caption = Caption + VectorMarker,
                .IsMoveable = True,
                .IsSelectable = True
            }
            displayObjects.Insert(1, displayObjectVector)
        End If
        Return displayObjects.ToArray()
    End Function

    Public Overrides Function Clone() As Object Implements ICloneable.Clone
        Return CloneWithNewPoints(Me.Points)
    End Function

    Public Overloads Function CloneWithNewPoints(points As PointF()) As RegionWithVector
        Dim obj = New RegionWithVector(points) With
        {
            .Caption = Me.Caption,
            .ID = Me.ID,
            .Color = Me.Color,
            .Vector = Me.Vector,
            .Description = Me.Description
        }
        For Each kvp In Me.Parameters.ToArray()
            obj.Parameters.Add(kvp.Key, kvp.Value)
        Next
        Return obj
    End Function

    Public Shared Function GetVectorLines(vector As Vector) As Line()
        Dim lines As New List(Of Line)
        Dim mulX = 1.0
        Dim mulY = 1.0
        Dim x1 = vector.Point1.X
        Dim y1 = vector.Point1.Y
        Dim x2 = vector.Point2.X
        Dim y2 = vector.Point2.Y
        Dim dx = x2 - x1
        Dim dy = y2 - y1
        Dim length = Math.Sqrt(dx ^ 2 + dy ^ 2)
        If length > 0 Then
            Dim angle As Single = CSng(Math.Atan2(dy, dx))
            Dim sz = 0.005
            Dim line1 = New Line(CSng(x1 * mulX + CSng(Math.Cos(angle - Math.PI / 2.0) * sz)), CSng(y1 * mulY + CSng(Math.Sin(angle - Math.PI / 2.0) * sz)),
                                 CSng(x2 * mulX), CSng(y2 * mulY))
            Dim line2 = New Line(CSng(x1 * mulX + CSng(Math.Cos(angle + Math.PI / 2.0) * sz)), CSng(y1 * mulY + CSng(Math.Sin(angle + Math.PI / 2.0) * sz)),
                                 CSng(x2 * mulX), CSng(y2 * mulY))
            Dim line3 = New Line(CSng(x1 * mulX + CSng(Math.Cos(angle - Math.PI / 2.0) * sz)), CSng(y1 * mulY + CSng(Math.Sin(angle - Math.PI / 2.0) * sz)),
                                 CSng(x1 * mulX + CSng(Math.Cos(angle + Math.PI / 2.0) * sz)), CSng(y1 * mulY + CSng(Math.Sin(angle + Math.PI / 2.0) * sz)))
            lines.AddRange({line1, line2, line3})
        End If
        Return lines.ToArray()
    End Function
End Class

File: Bwl.Imaging.Skia/Geometry/Region/RegionWithVector.cs
using SkiaSharp;

namespace Bwl.Imaging.Skia;


public class RegionWithVector : Region, ICloneable
{

    public static string VectorMarker { get; private set; } = "Vector";

    public string VectorCaption
    {
        get
        {
            return Caption + VectorMarker;
        }
    }

    public string VectorID
    {
        get
        {
            return ID + VectorMarker;
        }
    }

    public SKPoint Vector { get; set; } = new SKPoint();

    public Vector ImagingVector
    {
        get
        {
            var point1 = new SKPoint(Points.Average(item => item.X), Points.Average(item => item.Y));
            float point2X = point1.X + Vector.X;
            point2X = point2X < 0f ? 0f : point2X;
            point2X = point2X > 1f ? 1f : point2X;
            float point2Y = point1.Y + Vector.Y;
            point2Y = point2Y < 0f ? 0f : point2Y;
            point2Y = point2Y > 1f ? 1f : point2Y;
            var point2 = new SKPoint(point2X, point2Y);
            return new Vector(point1, point2);
        }
        set
        {
            Vector = new SKPoint(value.Points[1].X - value.Points[0].X, value.Points[1].Y - value.Points[0].Y);
        }
    }

    public RegionWithVector() : base(Guid.NewGuid().ToString(), new[] { new SKPoint(), new SKPoint(), new SKPoint(), new SKPoint() }) // Как Tetragon
    {
    }

    public RegionWithVector(SKPoint[] points) : base(Guid.NewGuid().ToString(), points)
    {
    }

    public RegionWithVector(string id) : base(id, new[] { new SKPoint(), new SKPoint(), new SKPoint(), new SKPoint() }) // Как Tetragon
    {
    }

    public RegionWithVector(string id, SKPoint[] points) : base(id, points) // Как Tetragon
    {
    }

    public override DisplayObject[] ToDisplayObjects(bool fullDisplay = true, float textSizeF = 0.013f, string channelIdxKey = "{ChannelIdxKey}")
    {
        var displayObjects = new List<DisplayObject>(base.ToDisplayObjects(fullDisplay, textSizeF, channelIdxKey));
        // Vector
        if (Vector.X != 0f | Vector.Y != 0f)
        {
            var displayObjectVector = new DisplayObject(VectorID, Color, ImagingVector)
            {
                Caption = Caption + VectorMarker,
                IsMoveable = true,
                IsSelectable = true
            };
            displayObjects.Insert(1, displayObjectVector);
        }
        return displayObjects.ToArray();
    }

    public override object Clone()
    {
        return CloneWithNewPoints(Points);
    }

    public new RegionWithVector CloneWithNewPoints(SKPoint[] points)
    {
        var obj = new RegionWithVector(points)
        {
            Caption = Caption,
            ID = ID,
            Color = Color,
            Vector = Vector,
            Description = Description
        };
        foreach (var kvp in Parameters.ToArray())
            obj.Parameters.Add(kvp.Key, kvp.Value);
        return obj;
    }

    public static Line[] GetVectorLines(Vector vector)
    {
        var lines = new List<Line>();
        double mulX = 1.0d;
        double mulY = 1.0d;
        float x1 = vector.Point1.X;
        float y1 = vector.Point1.Y;
        float x2 = vector.Point2.X;
        float y2 = vector.Point2.Y;
        float dx = x2 - x1;
        float dy = y2 - y1;
        double length = Math.Sqrt(Math.Pow((double)dx, 2d) + Math.Pow((double)dy, 2d));
        if (length > 0d)
        {
            float angle = (float)Math.Atan2((double)dy, (double)dx);
            double sz = 0.005d;
            var line1 = new Line((float)((double)x1 * mulX + (double)(float)(Math.Cos((double)angle - Math.PI / 2.0d) * sz)), (float)((double)y1 * mulY + (double)(float)(Math.Sin((double)angle - Math.PI / 2.0d) * sz)), (float)((double)x2 * mulX), (float)((double)y2 * mulY));
            var line2 = new Line((float)((double)x1 * mulX + (double)(float)(Math.Cos((double)angle + Math.PI / 2.0d) * sz)), (float)((double)y1 * mulY + (double)(float)(Math.Sin((double)angle + Math.PI / 2.0d) * sz)), (float)((double)x2 * mulX), (float)((double)y2 * mulY));
            var line3 = new Line((float)((double)x1 * mulX + (double)(float)(Math.Cos((double)angle - Math.PI / 2.0d) * sz)), (float)((double)y1 * mulY + (double)(float)(Math.Sin((double)angle - Math.PI / 2.0d) * sz)), (float)((double)x1 * mulX + (double)(float)(Math.Cos((double)angle + Math.PI / 2.0d) * sz)), (float)((double)y1 * mulY + (double)(float)(Math.Sin((double)angle + Math.PI / 2.0d) * sz)));
            lines.AddRange(new[] { line1, line2, line3 });
        }
        return lines.ToArray();
    }
}